[Woo POS] Make floating `Connect your reader` button start payment when tapped in checkout

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13431 
Merge after: #13505
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR addresses the unfinished parts of #13414 after #13505.

> Finally, the remaining thing to do is make payment start when you use the floating action button to connect the reader:
> 
> > The payment preparation process should start whether the small floating button is used, or the large purple button.
> > 
> > This should only happen when we're actually on the Checkout screen.

We achieve this by observing `TotalsViewModel.connectionStatus` while the `TotalsView` is visible and there isn't a reader connected. When a reader connects, we call `collectPayment`, having first waited for the order sync to complete.

We only observe the connected reader status while the `TotalsView` is visible, so we don't start any payments for connections started on the Product List.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app
2. Go to POS mode
3. Don't connect the reader, just add something to the cart
4. Tap Check out
5. On the Checkout screen, tap the floating `Connect your reader` button
6. Observe that when the connection completes, we prepare the reader for payment.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

There are a few criteria which affect how the merchant can get to a connected reader. 
- The purple CTA will show when the order's created, synced, and there's no reader connected. 
- The floating button always shows, and can be accessed from other screens
- We only listen to the reader connection status to start payments automatically (whichever button is used to start the connection) when the TotalsView is shown.
- But we shouldn't start that payment if the order hasn't synced at the moment the connection starts, even on `TotalsView` – in that case, delay until we finish syncing the order.
- And if the reader's already connected when the `TotalsView` shows, we should start the payment as soon as the order syncs

I've tested several scenarios here:

- [x] Instructions above
- [x] As above, but go back after first seeing the checkout, without connecting a reader, change the order, then complete the steps.
- [x] Breakpoint the Order Sync network request, and complete the connection before the order is synced – it should work when you let the breakpoint continue
- [x] Test with the big purple `Connect your reader` CTA that it still works correctly
- [x] Test that if you tap `Check out` with a reader connected, payment starts immediately.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/1de52515-d0d7-4f19-8eae-f58b37d02a5a


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [ ] This PR includes refactoring; smoke testing of the entire section is needed.

joshheald committed:
13431 Improve checkOutTapped naming and interface

This change stops us sharing the internal logic of the totalsViewModel through the name

joshheald committed:
13431 Make disconnected view action: connectReader

We will make the TotalsView automatically collect payment when it detects a connected reader, so the action is just to connect a reader now.

joshheald committed:
13431 Auto-collect payment when reader connect

This fixes the situation where connecting to a reader on the checkout screen using the floating button left the merchant without a clear next action, having to go back and forward again to trigger the payment.

joshheald committed:
Fix optional warning

joshheald committed:
13431 Add proper MockPOSOrderService

This keeps the preview service simple and for previews only, while allowing us to add more testing-related features to the proper mock.

joshheald committed:
13431 Test we collectPayment when reader connects

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
13431 Ensure we only start payment when on totals

By only listening for connectedReaders when TotalsView is on screen, we can avoid starting the payment if connection happens while it’s in the background.

When the merchant taps `Check out`, we’ll do a fresh check and start the payment if there’s a reader connected.

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/ViewModels/TotalsViewModelTests.swift:
```diff
@@ -23,7 +23,7 @@ final class TotalsViewModelTests: XCTestCase {
     }
     func test_isSyncingOrder() {}
-    func test_startSyncOrder() {}
+    func test_on_checkOutTapped_startSyncOrder() {}
```
Just linking for reference, we have a logged issue to fill these missing tests: https://github.com/woocommerce/woocommerce-ios/issues/13217

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/Mocks/MockPOSOrderService.swift:
```diff
@@ -0,0 +1,16 @@
+
+class MockPOSOrderService: POSOrderServiceProtocol {
+    var orderToReturn: Order?
+    func syncOrder(cart: [Yosemite.POSCartItem], order: Yosemite.Order?, allProducts: [any Yosemite.POSItem]) async throws -> Yosemite.Order {
```
Unrelated to this PR, it just came up because the return type of this mock and the unit tests.

I see we're using `OrderItem` and `Order` in several places within POS scope (`MockPOSOrderService`, `POSOrderPreviewService`, etc, ... ), rather than the previous POS versions (eg: `POSOrder`). The change was done [here](https://github.com/woocommerce/woocommerce-ios/pull/13381) when addressing tax calculation, but this change seems to break the differentiation between `app` and `pos`?

I have no specific considerations at the moment, just raising this up in case it is something we want to re-evaluate. I left a comment in that PR for the time being.

joshheald replied:
Yeah, I would prefer if we could nip this in the bud as well, and go back to using simpler objects in the POS code.

iamgabrielma commented on WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift:
```diff
@@ -106,26 +106,34 @@ final class TotalsViewModel: ObservableObject, TotalsViewModelProtocol {
-        Task { @MainActor in
-            await collectPayment()
+    func connectReaderTapped() {
+        Task { @MainActor [weak self] in
+            guard let self else {
+                return
+            }
+
```
Is this needed? We're not explicitly referencing self and the compiler is fine with removing it, but if we expect the task lifespan to be longer than self, then perhaps let's add a comment so it's not removed by mistake in the future. 

```suggestion
        Task { @MainActor  in
```

joshheald replied:
Hmm I didn't know we could do that. I wonder if it's something specific to `Task`, different to a normal closure.

We've got an implicit self here, so I thought that would only work if we strongly capture self.
It's `self.cardPresentPaymentService` – but you're absolutely right that it's not needed, so I'll make this change. 

Thanks for pointing it out.

iamgabrielma reviewed (approved):
LGTM! 🚢 

I've tested all cases above and marked the checkboxes 👍

joshheald committed:
13431 Remove unnecessary self capture
